llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp

```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    EntryProperties EntryProp = MDAnalysisInfo.EntryPropertyVec[0];
+    // ShaderFlagsAnalysis pass needs to collect and provide ShaderFlags for
+    // each entry function. Currently, even though the ShaderFlags value
+    // provided by ShaderFlagsAnalysis pass is created by walking all the
+    // function instructions of the module, it is sufficient to since there is
+    // only one entry function in the module.
+    ShaderEntryMDInfo SEP(EntryProp, Ctx, MDAnalysisInfo.ShaderProfile,
+                          MDResources, ShaderFlags);
+    MDTuple *MDT = SEP.getAsMetadata();
+    EntryFnMDNodes.emplace_back(MDT);
+  } break;
+  default:
+    assert(0 && "invalid profile");
+    break;
+  }
+
+  NamedMDNode *EntryPointsNamedMD =
+      M.getOrInsertNamedMetadata("dx.entryPoints");
+  for (auto *Entry : EntryFnMDNodes)
+    EntryPointsNamedMD->addOperand(Entry);
+}
+
 static void translateMetadata(Module &M, const DXILResourceMap &DRM,
                               const dxil::Resources &MDResources,
-                              const ComputedShaderFlags &ShaderFlags) {
-  dxil::ValidatorVersionMD ValVerMD(M);
-  if (ValVerMD.isEmpty())
-    ValVerMD.update(VersionTuple(1, 0));
-  dxil::createShaderModelMD(M);
-  dxil::createDXILVersionMD(M);
+                              const ComputedShaderFlags &ShaderFlags,
+                              const dxil::ModuleMetadataInfo &MDAnalysisInfo) {
+  LLVMContext &Ctx = M.getContext();
+  IRBuilder<> IRB(Ctx);
+  if (MDAnalysisInfo.ValidatorVersion.empty()) {
+    // Module has no metadata node signifying valid validator version.
+    // Create metadata dx.valver node with version value of 1.0
+    const VersionTuple DefaultValidatorVer{1, 0};
+    Metadata *MDVals[2];
+    MDVals[0] =
+        ConstantAsMetadata::get(IRB.getInt32(DefaultValidatorVer.getMajor()));
+    MDVals[1] = ConstantAsMetadata::get(
+        IRB.getInt32(DefaultValidatorVer.getMinor().value_or(0)));
+    NamedMDNode *ValVerNode = M.getOrInsertNamedMetadata("dx.valver");
+    ValVerNode->addOperand(MDNode::get(Ctx, MDVals));
+  }
+
+  Metadata *SMVals[3];
+  VersionTuple SM = MDAnalysisInfo.ShaderModelVersion;
+  SMVals[0] =
+      MDString::get(Ctx, getShortShaderStage(MDAnalysisInfo.ShaderProfile));
+  SMVals[1] = ConstantAsMetadata::get(IRB.getInt32(SM.getMajor()));
+  SMVals[2] = ConstantAsMetadata::get(IRB.getInt32(SM.getMinor().value_or(0)));
+  NamedMDNode *SMMDNode = M.getOrInsertNamedMetadata("dx.shaderModel");
+  SMMDNode->addOperand(MDNode::get(Ctx, SMVals));
+
+  VersionTuple DXILVer = MDAnalysisInfo.DXILVersion;
+  Metadata *DXILVals[2];
+  DXILVals[0] = ConstantAsMetadata::get(IRB.getInt32(DXILVer.getMajor()));
+  DXILVals[1] =
+      ConstantAsMetadata::get(IRB.getInt32(DXILVer.getMinor().value_or(0)));
+  NamedMDNode *DXILVerMDNode = M.getOrInsertNamedMetadata("dx.version");
+  DXILVerMDNode->addOperand(MDNode::get(Ctx, DXILVals));
```

Comment:
I'd probably move these three blocks of logic into simple functions to keep the level of abstraction at a single level here, so this function would just look something like:
```c++
  emitValidatorVersionMD(M, MMI);
  emitShaderModuleMD(M, MMI);
  emitDXILVersionMD(M, MMI);
  NamedMDNode *ResourceNode = emitResourceMetadata(M, DRM, MDResources);
  
  SmallVector<MDNode *> EntryNodes;
  if (MMI.ShaderProfile == Triple::EnvironmentType::Library)
    EntryNodes.push_back(emitTopLevelLibraryNode(M, ResourceNode, ShaderFlags));
  else if (MMI.EntryProps.size() != 1)
    diagnoseWrongNumberOfEntryProps(MMI.ShaderProfile, MMI.EntryProps.size());
  for (const EntryProperties &Entry : MMI.EntryProps)
    EntryNodes.push_back(emitEntryNode(M, Entry, MMI, ResourceNode, ShaderFlags));
  emitEntryPointsMD(M, EntryNodes);
```

Reply:
> I'd probably move these three blocks of logic into simple functions to keep the level of abstraction at a single level here, so this function would just look something like:
> 
> ```c++
>   emitValidatorVersionMD(M, MMI);
>   emitShaderModuleMD(M, MMI);
>   emitDXILVersionMD(M, MMI);
>   NamedMDNode *ResourceNode = emitResourceMetadata(M, DRM, MDResources);
>   
>   SmallVector<MDNode *> EntryNodes;
>   if (MMI.ShaderProfile == Triple::EnvironmentType::Library)
>     EntryNodes.push_back(emitTopLevelLibraryNode(M, ResourceNode, ShaderFlags));
>   else if (MMI.EntryProps.size() != 1)
>     diagnoseWrongNumberOfEntryProps(MMI.ShaderProfile, MMI.EntryProps.size());
>   for (const EntryProperties &Entry : MMI.EntryProps)
>     EntryNodes.push_back(emitEntryNode(M, Entry, MMI, ResourceNode, ShaderFlags));
>   emitEntryPointsMD(M, EntryNodes);
> ```

Done.